docs/okeps/okep-5224-connecting-udns/okep-5224-connecting-udns.md

```diff
@@ -0,0 +1,1652 @@
+networks to be connected together.
+
+`ClusterNetworkConnect` CR will be cluster-scoped resource and admin-only
+API that will allow admins to select multiple different networks that
+need to be connected together. It will also allow specifying what level
+of connectivity is expected. See the API definitions for more details.
+
+Admins can create a `ClusterNetworkConnect` CR which signals the
+controller to connect the topologies of the UDNs selected by this CR.
+
+Sample YAML:
+
+```yaml
+apiVersion: k8s.ovn.org/v1
+kind: ClusterNetworkConnect
+metadata:
+  name: colored-enterprise
+spec:
+  networkSelectors: # can match on UDNs and/or CUDNs
+    - networkSelectionType: ClusterUserDefinedNetworks
+      clusterUserDefinedNetworkSelector:
+        networkSelector:
+          matchExpressions:
+          - key: kubernetes.io/metadata.name
+            operator: In
+            values:
+            - blue-network
+            - green-network
+    - networkSelectionType: PrimaryUserDefinedNetworks
+      primaryUserDefinedNetworkSelector:
+        namespaceSelector:
+          matchExpressions:
+          - key: kubernetes.io/metadata.name
+            operator: In
+            values:
+            - yellow
+            - red
+  connectSubnets: # can have at most 1 CIDR for each family type
+  - cidr: "192.168.0.0/16"  # 65K total IPs
+    layer3NetworkPrefix: 24 # 255 IPs per layer3 network = ~100 possible nodes in your cluster
+  - cidr: "fd01::/64"
+    layer3NetworkPrefix: 96
+  connectivityEnabled:
+    - PodNetwork
+    - ClusterIPServiceNetwork
+```
+* `connectivityEnabled` field will ensure we can add support for enabling
+  more types of features for the connected UDNs granularly in the future.
+  At least one of these options must be set and validations will be added
+  for which combinations of these options could/should be set together
+* If `PodNetwork` is enabled then admin is asking for full
+  pod2pod connectivity across UDNs
+* If `ClusterIPServiceNetwork` is enabled then admin is asking for
+  clusterIPs to be reachable across UDNs.
+    * If this flag is set but `PodNetwork` is not set, then
+      it means users only want partial connectivity through services and
+      not full mesh pod connectivity.
+* `connectSubnets` field is used to take a configurable subnet from the end
+  user that will be used to connect the different networks together. The
+  slice needs to be big enough to accommodate an IP per network per node
+  for layer3 network connectivity. So if there are M nodes in your cluster
+  and N_L3 networks you should have N_L3×M IPs for allocation in this slice.
+  The `layer3NetworkPrefix` contains the prefix length allocated for each
+  connected layer3 network. From this pool each node get's a /31 point to
+  point network for actual topology connectivity.
+  To connect layer2 networks we don't need per node-network slices, we just
+  need one `/31` slice per network. Hence we don't need a `layer2NetworkPrefix`.
```

Comment:
It is directly related though to "layer3NetworkPrefix" because it still has to carve a subnet from those bits. I wouldn't call it layer3NetworkPrefix, I would just call it NetworkPrefix, or NetworkMask. We still want to make it a required field, because we want the user to have to be aware of this setting. I think they can put in a value that is the same prefix as the cidr, and then that would signal that we should not carve any network specific subnerts, and just use the range for /31s on Layer2. If a Layer3 is selected, then we report error status.

Reply:
I guess that's true! cause I do carve out the layer3NetworkPrefix first even for layer2 networks. I'll rename this to NetworkPrefix then as initially agreed.